File: flutter_inappwebview_forge_android/android/src/main/kotlin/com/emirkanacar/flutter_inappwebview_forge_android/webview/web_message/WebMessageChannelChannelDelegate.kt
package com.emirkanacar.flutter_inappwebview_forge_android.webview.web_message

import com.emirkanacar.flutter_inappwebview_forge_android.types.ChannelDelegateImpl
import com.emirkanacar.flutter_inappwebview_forge_android.types.WebMessageCompatExt
import com.emirkanacar.flutter_inappwebview_forge_android.webview.in_app_webview.InAppWebView
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.HashMap

open class WebMessageChannelChannelDelegate(
    private var webMessageChannel: WebMessageChannel?,
    channel: MethodChannel
) : ChannelDelegateImpl(channel) {
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "setWebMessageCallback" -> {
                val channel = webMessageChannel
                if (channel?.webView is InAppWebView) {
                    val index = call.argument<Int>("index")
                    if (index == null) {
                        result.error("invalid_arguments", "Web message port index is required.", null)
                    } else {
                        channel.setWebMessageCallbackForInAppWebView(index, result)
                    }
                } else {
                    result.success(false)
                }
            }

            "postMessage" -> {
                val channel = webMessageChannel
                if (channel?.webView is InAppWebView) {
                    val index = call.argument<Int>("index")
                    val message = WebMessageCompatExt.fromMap(
                        call.argument<MutableMap<String, Any?>>("message")
                    )
                    if (index == null || message == null) {
                        result.error("invalid_arguments", "Web message index and message are required.", null)
                    } else {
                        channel.postMessageForInAppWebView(index, message, result)
                    }
                } else {
                    result.success(false)
                }
            }

            "close" -> {
                val channel = webMessageChannel
                if (channel?.webView is InAppWebView) {
                    val index = call.argument<Int>("index")
                    if (index == null) {
                        result.error("invalid_arguments", "Web message port index is required.", null)
                    } else {
                        channel.closeForInAppWebView(index, result)
                    }
                } else {
                    result.success(false)
                }
            }

            else -> result.notImplemented()
        }
    }

    open fun onMessage(index: Int, message: WebMessageCompatExt?) {
        val channel = getChannel() ?: return
        val obj = HashMap<String, Any?>()
        obj["index"] = index
        obj["message"] = message?.toMap()
        channel.invokeMethod("onMessage", obj)
    }

    override fun dispose() {
        super.dispose()
        webMessageChannel = null
    }
}

File: flutter_inappwebview_forge_ios/ios/flutter_inappwebview_forge_ios/Sources/flutter_inappwebview_forge_ios/InAppWebView/WebMessage/WebMessageChannelChannelDelegate.swift
//
//  WebMessageChannelChannelDelegate.swift
//  flutter_inappwebview
//
//  Created by Lorenzo Pichilli on 07/05/22.
//

import Foundation
import Flutter

public class WebMessageChannelChannelDelegate: ChannelDelegate {
    private weak var webMessageChannel: WebMessageChannel?
    private var pendingResults: [UUID: FlutterResult] = [:]
    
    public init(webMessageChannel: WebMessageChannel, channel: FlutterMethodChannel) {
        super.init(channel: channel)
        self.webMessageChannel = webMessageChannel
    }
    
    public override func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard canDispatchCallbacks() else {
            result(nil)
            return
        }
        let arguments = call.arguments as? NSDictionary
        
        switch call.method {
        case "setWebMessageCallback":
            if let _ = webMessageChannel?.webView, let ports = webMessageChannel?.ports, ports.count > 0 {
                guard let index = arguments?["index"] as? Int, ports.indices.contains(index) else {
                    result(FlutterError(code: "invalid_arguments", message: "Invalid port index.", details: nil))
                    return
                }
                let port = ports[index]
                let requestId = track(result)
                do {
                    try port.setWebMessageCallback { [weak self] (_) in
                        self?.complete(requestId, value: true)
                    }
                } catch let error as NSError {
                    complete(requestId, value: FlutterError(code: "WebMessageChannel", message: error.domain, details: nil))
                }
                
            } else {
                result(true)
            }
            break
        case "postMessage":
            if let webView = webMessageChannel?.webView, let ports = webMessageChannel?.ports, ports.count > 0 {
                guard let index = arguments?["index"] as? Int,
                      ports.indices.contains(index),
                      let messageMap = arguments?["message"] as? [String: Any?] else {
                    result(FlutterError(code: "invalid_arguments", message: "Invalid port message.", details: nil))
                    return
                }
                let message = WebMessage.fromMap(map: messageMap)
                let port = ports[index]
                
                var ports: [WebMessagePort] = []
                if let notConnectedPorts = message.ports {
                    for notConnectedPort in notConnectedPorts {
                        if let webMessageChannel = webView.webMessageChannels[notConnectedPort.webMessageChannelId] {
                            ports.append(webMessageChannel.ports[Int(notConnectedPort.index)])
                        }
                    }
                }
                message.ports = ports

                let requestId = track(result)
                do {
                    try port.postMessage(message: message) { [weak self] (_) in
                        self?.complete(requestId, value: true)
                    }
                } catch let error as NSError {
                    complete(requestId, value: FlutterError(code: "WebMessageChannel", message: error.domain, details: nil))
                }
            } else {
                result(true)
            }
            break
        case "close":
            if let _ = webMessageChannel?.webView, let ports = webMessageChannel?.ports, ports.count > 0 {
                guard let index = arguments?["index"] as? Int, ports.indices.contains(index) else {
                    result(FlutterError(code: "invalid_arguments", message: "Invalid port index.", details: nil))
                    return
                }
                let port = ports[index]
                let requestId = track(result)
                do {
                    try port.close { [weak self] (_) in
                        self?.complete(requestId, value: true)
                    }
                } catch let error as NSError {
                    complete(requestId, value: FlutterError(code: "WebMessageChannel", message: error.domain, details: nil))
                }
            } else {
                result(true)
            }
            break
        default:
            result(FlutterMethodNotImplemented)
            break
        }
    }
    
    public func onMessage(index: Int64, message: WebMessage?) {
        guard canDispatchCallbacks() else { return }
        let arguments: [String:Any?] = [
            "index": index,
            "message": message?.toMap()
        ]
        channel?.invokeMethod("onMessage", arguments: arguments)
    }
    
    public override func dispose() {
        let results = pendingResults.values
        pendingResults.removeAll()
        results.forEach { $0(nil) }
        super.dispose()
        webMessageChannel = nil
    }

    private func canDispatchCallbacks() -> Bool {
        guard channel != nil, let webMessageChannel = webMessageChannel else { return false }
        guard let webView = webMessageChannel.webView else { return true }
        return webView.acceptsCallbacks()
    }

    private func track(_ result: @escaping FlutterResult) -> UUID {
        let requestId = UUID()
        pendingResults[requestId] = result
        return requestId
    }

    private func complete(_ requestId: UUID, value: Any?) {
        guard let result = pendingResults.removeValue(forKey: requestId) else { return }
        result(value)
    }
    
    deinit {
        dispose()
    }
}
